pennylane/math/quantum.py

```diff
@@ -606,3 +606,158 @@ def _compute_mutual_info(
     )
 
     return vn_entropy_1 + vn_entropy_2 - vn_entropy_12
+
+
+def fidelity(state0, state1, check_state=False, c_dtype="complex128"):
+    r"""Compute the fidelity for two states (a state can be a state vector or a density matrix) acting on quantum
+    systems with the same size. For two pure states, the fidelity corresponds to the squared overlap. For a pure state
+    and a mixed state, it corresponds to the squared expectation of the mixed state in the pure state. Finally for two
+    mixed states, it is defined by the last formula:
+
+    .. math::
+        F( \ket{\psi} , \ket{\phi}) = \left|\bra{\psi}\ket{\phi}\right|^2
+
+        F( \ket{\psi} , \sigma ) = \left|\bra{\psi} \sigma \ket{\psi}\right|^2
+
+        F( \rho , \sigma ) = -\text{Tr}( \sqrt{\sqrt{\rho} \sigma \sqrt{\rho}})^2
+
+    .. warning::
+        The second state is coerced to the type and dtype of the first state. The fidelity is returned in the type
+        of the interface of the first state.
+
+    Args:
+        state0 (tensor_like): 1D state vector or 2D density matrix
+        state1 (tensor_like): 1D state vector or 2D density matrix
+        check_state (bool): If True, the function will check the state validity (shape, norm) or (shape, trace,
+            semi positive definitiveness).
+        c_dtype (str): Complex floating point precision type.
+
+    Returns:
+        float: Fidelity between the two quantum states.
+
+    **Example**
+
+        >>> state0 = [0.98753537-0.14925137j, 0.00746879-0.04941796j]
+        >>> state1 = [0.99500417+0.j, 0.09983342+0.j]
+        >>> qml.math.fidelity(state0, state1)
+        0.9905158135644924
+
+        >>> state0 = [0, 1]
+        >>> state1 = [[0, 0], [0, 1]]
+        >>> qml.math.fidelity(state0, state1)
+        1.0
+
+        >>> state0 = [[1, 0], [0, 0]]
+        >>> state1 = [[0, 0], [0, 1]]
+        >>> qml.math.fidelity(state0, state1)
+        0.0
+
+    """
+    # Cast as a c_dtype array
+    state0 = cast(state0, dtype=c_dtype)

```

Comment:
The description says it is casting to the dtype of the first state but here you are fixing it externally?